### DISCUSSION TITLE
ValueError: cannot reshape array of size 1024 into shape (1920,1080,newaxis)

### DISCUSSION THREAD

**Discussion by sugizo** [streamlink/streamlink#5912] [Q&A] [answered]
### Description

# objective
write output in python API

## trial using vidgear writer
**steps**
`pip install -U vidgear streamlink
`
**code**
```
import streamlink
import numpy as np
from vidgear.gears import WriteGear
output = 'rtsp_video_capture_stream.mp4'
cap = streams["best"].open()
output_params = {"-vf": "scale = 640 : -1"}
writer = WriteGear(output = output, compression_mode = True, logging = False, **output_params)
while True: 
    frame = cap.read(1024)
    writer = np.frombuffer(frame, np.uint8).reshape(1920, 1080, -1)
    vidout.write(image)
cap.release()
writer.close()
```

**result**
```
ValueError                                Traceback (most recent call last)
[<ipython-input-46-93012a2a3f72>](https://localhost:8080/#) in <cell line: 8>()
      8 while True:
      9     frame = cap.read(1024)
---> 10     writer = np.frombuffer(frame, np.uint8).reshape(1920, 1080, -1)
     11     vidout.write(image)
     12 cap.release()

ValueError: cannot reshape array of size 1024 into shape (1920,1080,newaxis)
```

## trial using ffmpegcv writer
**steps**
`pip install -U ffmpegcv streamlink`

**code**
```
import streamlink
import ffmpegcv
import numpy as np
output_file = 'rtsp_video_capture_stream.mp4'
cap = streams["best"].open()
vidout = ffmpegcv.VideoWriter(output_file, 'hevc', pix_fmt = 'bgr24')
while True: 
    frame = cap.read(1024)
    image = np.frombuffer(frame, np.uint8).reshape(1920, 1080, -1)
    vidout.write(image)
cap.release()
vidout.release()
```

**result**
```
ValueError                                Traceback (most recent call last)
[<ipython-input-47-2ea4dd9b8952>](https://localhost:8080/#) in <cell line: 7>()
      7 while True:
      8     frame = cap.read(1024)
----> 9     image = np.frombuffer(frame, np.uint8).reshape(1920, 1080, -1)
     10     vidout.write(image)
     11 cap.release()

ValueError: cannot reshape array of size 1024 into shape (1920,1080,newaxis)``

**Comment by bastimeyer** [comment, marked as answer]
What you're trying to do doesn't make any sense.

You're reading 1024 bytes of Streamlink's output, which could be any kind of data like MPEG-2 transport stream data, MPEG-4 stream data or literally anything else, and then you're trying to put these 1024 bytes into a `numpy.ndarray` which you're then trying to [reshape](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html#numpy.reshape) into a multi-dimensional array with dimension values which resemble the numbers of a common video resolution.

This is not how video data works. This is also very much unrelated to Streamlink. Not to mention that you're not even asking a question and just posting random code and the error message.

What you're doing is essentially this

```py
import numpy as np
np.frombuffer(bytes(1024), np.uint8).reshape(1920, 1080, -1)
```

which as you can see, is totally unrelated to Streamlink, and even if it were, it doesn't make any sense in terms of media streams.

What you want is to demux the stream data first, then decode (one of) the video stream(s) and get the pixel data of a specific frame. This is out of the scope of this discussion forum and you'll have to ask somewhere else.

Good luck.
